### ISSUE TITLE
Make PutBlock<T>() idempotent

### ISSUE THREAD

**Issue by earlbread** [planetarium/libplanet#616]
When we call `LiteDBStore.PutBlock<T>()` with a block that already exists, the block is rewritten. However, if a block already exists, doing nothing is better for performance.

**Comment by mahditakrim** [comment]
Hi @earlbread ,
I agree with you (good performance lover ;D) .
May I take a look ?

**Comment by earlbread** [comment]
Sure, thanks! @mahditakrim

**Comment by mahditakrim** [comment]
Dear @earlbread ,
I just sent a pull request but 1 check failed , it's changelog.

**Comment by mahditakrim** [comment]
Dear @earlbread ,
I took a look at the codes, and I got questions:

1. How to know that a block is already exists, I mean I know that blocks got unique hash codes so how to compare a block hash code to other hash codes (if I'm right) ?

2. Is it better to have a method which can determine that a block exists **or** just put the if statement in the usage code ?

**Comment by earlbread** [comment]
1. You can check whether a block already exists by using `_db.FileStorage.FindById()` method and `BlockFileId()` method. Please refer to this commit 19c613241a93a3c5d63a768bd3f78ed900bbf6ab.

2. I think the `if` statement is enough.

3. To build and test this project, please refer to this document. https://github.com/planetarium/libplanet/blob/master/CONTRIBUTING.md

**Comment by mahditakrim** [comment]
> 1. You can check whether a block already exists by using `_db.FileStorage.FindById()` method and `BlockFileId()` method. Please refer to this commit [19c6132](https://github.com/planetarium/libplanet/commit/19c613241a93a3c5d63a768bd3f78ed900bbf6ab).
> 2. I think the `if` statement is enough.
> 3. To build and test this project, please refer to this document. https://github.com/planetarium/libplanet/blob/master/CONTRIBUTING.md
> 
> Please let me know if you have any questions.

Thanks

**Comment by earlbread** [comment]
`FileStorage.FindById` is from LiteDB(https://github.com/mbdavid/LiteDB/wiki/FileStorage). If the file exists, it returns `LiteFileInfo`, and if it is not `null`, we don't need to write the block again.

See `LiteDBStore.PutTransaction` method to know how to use `_db.FileStorage.FindById() ` method.(https://github.com/planetarium/libplanet/blob/master/Libplanet/Store/LiteDBStore.cs#L315).

**Comment by mahditakrim** [comment]
> `FileStorage.FindById` is from LiteDB(https://github.com/mbdavid/LiteDB/wiki/FileStorage). If the file exists, it returns `LiteFileInfo`, and if it is not `null`, we don't need to write the block again.
> 

You mean the block exists by file exists ?

**Comment by earlbread** [comment]
We have to return the method when the block is *not null* which means the block file already exists. So the code should be:

```
if (!(_db.FileStorage.FindById(BlockFileId(block.Hash)) is null))
{
    return;
}
```

**Comment by mahditakrim** [comment]
@earlbread ,
Got it, thanks for your patience.

**Comment by earlbread** [comment]
Closed by #647
